.github/workflows/pull_request.yaml

```
@@ -22,7 +24,7 @@ jobs:
         run: pip install pipenv==2023.12.1
       - name: Cache virtualenv
         id: cache-virtualenv
-        uses: actions/cache@v2
+        uses: actions/cache@v4
```

```suggestion
        uses: actions/cache@1bd1e32a3bdc45362d1e726936510720a7c30a57 # v4.2.0
```
As well as bumping, I'd suggest using the commit hash as per the recent recommendation from Github. There's also another one in a subsequent job that needs the same.

This bump to v4 does seem to have resolved the deprecation warning though 👍

---

Sounds good. As CodeQL only suggets this change for third party actions, i shall update the PR description for this. change 👍